File: C-Sharp/Reapply_pending_job/Reapply_pending_job.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using RestSharp;
using RestSharp.Authenticators;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;

namespace Reapply_pending_job
{
    class Reapply_job
    {
        public static string baseurl = "https://suremdm.42gears.com/api";
        private static string Username = "sa";
        private static string Password = "0000";
        private static string ApiKey = "37A007A7-5D12-4D14-974E-B211F8F378EA";
        static void Main(string[] args)
        {
            string DeviceID = GetDeviceID("Client04");
            if (DeviceID != null)
            {
                string status = Jobqueue(DeviceID, false);
                Console.WriteLine(status);
            }
            else
            {
                Console.WriteLine("Device not found!");
            }
            Console.ReadKey();
        }

        // Function for retrieving jobqueue of device
        private static string Jobqueue(string deviceID, Boolean bShowAll)
        {
            /* Retrieve Jobqueue of the device
               Endpoint: "/jobqueue/{deviceID}/{bShowAll}
                 Method: GET
                 Path parameters:
                    deviceID: Id of the device
                    bShowAll: boolean
                 Authentication:
                     Basic authentication
                 Headers:
                     ApiKey: “Your Api-Key”
             */

            string URL = baseurl + "/jobqueue/" + deviceID + "/" + bShowAll;
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            // Set request method
            var request = new RestRequest(Method.GET);
            // Execute method
            IRestResponse response = client.Execute(request);
            if (response.StatusCode == HttpStatusCode.OK)
            {
                List<JObject> jsonResponse = JsonConvert.DeserializeObject<List<JObject>>(response.Content);
                if ((jsonResponse != null) && (jsonResponse.Any()))
                {
                    foreach (var status in jsonResponse)
                    {
                        if (status.GetValue("Status").ToString() == "ERROR" || status.GetValue("Status").ToString() == "FAILED" || status.GetValue("Status").ToString() == "SCHEDULED")
                        {
                            string rowID = status.GetValue("RowId").ToString();
                            string jobID = status.GetValue("JobID").ToString();
                            Console.WriteLine(ReapplyJob(deviceID, jobID, rowID));
                        }
                    }
                    return "Success";
                }
            }
            return "Failed";
        }

        // Function to re-apply all the pending jobs
        private static string ReapplyJob(string deviceID, string jobID, string rowID)
        {
            /* Reapply job
               Endpoint: "/jobqueue/{jobID}/{deviceID}/{rowID};
                 Method: PUT
                 Path parameters:
                    jobID: ID of the job to apply
                    deviceID: Id of the device
                    rowID: Jobqueue ID of the job
                 Authentication:
                     Basic authentication
                 Headers:
                     ApiKey: “Your Api-Key”
             */

            string URL = baseurl + "/jobqueue/" + jobID + "/" + deviceID + "/" + rowID;
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            // Set request method
            var request = new RestRequest(Method.PUT);
            // Execute request
            IRestResponse response = client.Execute(request);
            return response.Content.ToString();
        }

        // Function to retrieve ID of the device
        private static string GetDeviceID(string deviceName)
        {
            /*  Retreiving information of device
                 Endpoint: /device:
                 Method: POST
                 Request Body:
                     {  
                        "ID": string,
	                    "SearchValue":string,
	                    "Limit":integer,
	                    "SortColumn":string,
	                    "SortOrder":string,
	                    "IsSearch":boolean,
	                    "SearchColumns":string[]
                    }
                 Authentication:
                     Basic authentication
                 Headers:
                     ApiKey: “Your Api-Key” 
             */

            string URL = baseurl + "/device";
            var client = new RestClient(URL);
            //  Basic authentication header
            client.Authenticator = new HttpBasicAuthenticator(Username, Password);
            //  ApiKey Header
            client.AddDefaultHeader("ApiKey", ApiKey);
            //  Set content type
            client.AddDefaultHeader("Content-Type", "application/json");
            //  Set request method
            var request = new RestRequest(Method.POST);
            //  Request payload
            var RequestPayLoad = new
            {
                ID = "AllDevices",
                IsSearch = true,
                Limit = 10,
                SearchColumns = new string[] { "DeviceName" },
                SearchValue = deviceName,
                SortColumn = "LastTimeStamp",
                SortOrder = "asc"
            };
            request.AddJsonBody(RequestPayLoad);

            //  Execute request
            IRestResponse response = client.Execute(request);

            if (response.StatusCode == HttpStatusCode.OK)
            {
                var OutPut = JsonConvert.DeserializeObject<JObject>(response.Content);
                foreach (var device in OutPut["rows"])
                {
                    if ((string)device["DeviceName"] == deviceName)
                    {
                        return device["DeviceID"].ToString();
                    }
                }
            }
            return null;
        }
    }
}

File: Python/Reapply_pending_job.py
# This program re-applies all pending jobs from device job queue
import requests,json

baseurl="https://suremdm.42gears.com/api"  # BaseURL of SureMDM
Username="Username"
Password="Password"
ApiKey="Your ApiKey"

# Method for getting device info
def GetDeviceID(deviceName):
    '''
    Retreiving information of device
        Endpoint: /device
        Method: POST
        Request Body:
            {  
                "ID": string,
                "SearchValue":string,
                "Limit":integer,
                "SortColumn":string,
                "SortOrder":string,
                "IsSearch":boolean,
                "SearchColumns":string[]
            }
        Authentication:
            Basic authentication
        Headers:
            ApiKey: “Your Api-Key”
    '''
    # Api url
    url = baseurl+"/device"
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Request body
    PayLoad={
        "ID" : "AllDevices",
        "IsSearch" : bool(1),
        "Limit" : 10,
        "SearchColumns" : ["DeviceName"],
        "SearchValue" : deviceName,
        "SortColumn" : "LastTimeStamp",
        "SortOrder" : "asc"
    }
    # Executing request
    response = requests.post(url,auth=Credentials,json=PayLoad,headers=headers)
    # Extracting required GroupID
    if response.status_code == 200:
        if response.text != '[]':
            data = response.json()
            for device in data['rows']:
                if device['DeviceName'] == deviceName:
                    return device["DeviceID"]
    else:
        return None


def ReapplyJob(deviceID,jobID,rowID):
    # Api url
    url = baseurl + "/jobqueue/" + jobID + "/" + deviceID + "/" + rowID
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Executing request
    response = requests.put(url,auth=Credentials,headers=headers)
    print(response.text)
    

def JobQueue(deviceID,bShowAll):
    '''
    Retrieve Jobqueue of the device
        Endpoint: "/jobqueue/{deviceID}/{bShowAll}
            Method: POST
            Path parameters:
            deviceID: Id of the device
            bShowAll: boolean
            Authentication:
                Basic authentication
            Headers:
                ApiKey: “Your Api-Key”
    '''
    # Api url
    url = baseurl + "/jobqueue/" + deviceID + "/" + str(bShowAll)
    # Headers
    headers = {
        # Api-Key header
        'ApiKey': ApiKey,
        # Set Content type
        'Content-Type': "application/json",
        }
    # Basic authentication credentials
    Credentials=(Username,Password)
    # Executing request
    response = requests.get(url,auth=Credentials,headers=headers)
    if response.status_code == 200:
        if response.text != '[]':
            data = response.json()
            for job in data:
                if job['Status'] == "ERROR" or job['Status'] == "FAILED" or job['Status']== "SCHEDULED":
                    rowID=job['RowId']
                    jobID=job['JobID']
                    ReapplyJob(deviceID,jobID,rowID)
                    return 'Success'
    else:
        return 'Failed'

# Main
DeviceID=GetDeviceID("Device_Name")
if DeviceID!=None:
    status=JobQueue(DeviceID,False)
    print(status)
else:
    print('Device not found')